ansible-doc: show 'notes', 'seealso', and top-level 'version_added' for role entrypoints

felixfontein opened the pull request:
##### SUMMARY
Fixes parts of #81735.

##### ISSUE TYPE
- Docs Pull Request
- Feature Pull Request

felixfontein committed:
Show 'notes', 'seealso', 'requirements', and top-level 'version_added' for role entrypoints.
Re-add 'attributes' for role entrypoints.

sivel commented:
cc @cidrblock @ssbarnea just an FYI about potential schema changes to role arg specs.  This hasn't been merged, but wanted to let you know ahead of time.

felixfontein commented:
This PR currently adds back role attributes (i.e. it fixes #82639).

bcoca commented:
i see the value of  `notes`, `seealso`  and `version_added`, but `requirements` and `attributes` per role entry point is not something i think we should incentivize.

felixfontein commented:
May I ask why? I think both `requirements` and `attributes` make perfect sense. I gave an example of attributes in https://github.com/ansible/ansible/issues/82639#issuecomment-1923197126. And roles also have requirements, similar to plugins and modules, so having these listed in a standard `Requirements` section instead of sometimes in `description`, sometimes in `notes`, and sometimes maybe somewhere else, is IMO a good idea.

bcoca commented:
requirements and attributes seem to be copies of those already existing for modules/plugins, roles themselves should not create unique ones, specially at the spec level.

I would much prefer being able to see an aggregate of the used plugins requirements/attributes than having yet another facility to list them that is actually detached from the code that is tied to them, leading to double bookeeping and all the issues that entails.

felixfontein commented:
> requirements and attributes seem to be copies of those already existing for modules/plugins

That's very much not the case. For example, a role administering some service or device using the `uri` module has more requirements than `uri`'s requirements. And whether a role is idempotent is totally unrelated to whether the modules/plugins used by that role are idempotent.

bcoca commented:
I'm not sure what idempotence has to do with this. As for requirements I would suggest that they are set at the role level, not at the role spec level.

felixfontein commented:
> I'm not sure what idempotence has to do with this.

```yaml
argument_specs:
  main:
    short_description: Set up GitLab
    ...
    attributes:
      idempotence:
        description: Whether the role is idempotent.
        support: partial
        details: If O(upgrade_to_latest=true) the role is not idempotent.
```

> As for requirements I would suggest that they are set at the role level, not at the role spec level.

Different entrypoints can have different requirements.

bcoca commented:
requirements: they can, that should not specify this at the entrypoint level, you would want those at the role level for installation

idempotence was not really  something i would express as an attribute. Also i would  not want to overload the 'attribute' facility  to be confused with the module/plugin specific ones, which will have programmatic ties in the future.

briantist commented:
I will add that I tried to use `requirements` recently and was surprised that it didn't work, so I said, "ok, I'll just put it in `notes:`" and then when that didn't work I gave up and added it to the `description:`.

I definitely would use these fields in roles, internally I have a bunch of collections and they contain mostly or exclusively roles; the documentation support for these is very important.

I also completely disagree that these fields should not change per entrypoint. I would _like_ to be able to specify them at the role level for roles where they are the same on every entrypoint, but I do have some roles where the requirements and especially notes would differ between them, so that option is also important (otherwise I will be back to overloading `description`).

I also agree with Felix about attributes: they are absolutely important, and again they should be per-entrypoint. The most useful attribute I can think of off the top of my head is around check mode support, but I see no reason to offer restrictions on them for roles specifically. 

> idempotence was not really something i would express as an attribute
I'm not sure why, it feels like a natural fit to me.. I would use an idempotence attribute since that's very important information, and it's "enforced" (somewhat) in my own roles via molecule's idempotence test in my CI, but it sounds like this whole concept might be a separate conversation and I don't know if we should derail the discussion of attributes being usable in roles with discussion of this specific attribute idea.

bcoca commented:
Im' not saying to not allow requirements, I'm saying they make more sense in meta/main.yml than in the role spec, as this is something to inspect on installation.  I reiterate the issue with 'double bookeeping' from modules/plugins and the distance between spec and task lists which also contributes to divergence. All of it made worse per spec if there is any overlap between them.  I had been toying with a 'role inspector' that would try/attempt to get that information from modules/plugins in use in the role itself. 

I understand why you want requirements/attributes per spec, but it goes against the intended design to keep roles small and to task. When we add infrastructure that is only useful for hugely complex and mixed used roles, we are operating against our own intent, we already regret adding `tasks/vars/etc_from` .

As for attributes for roles , this is not something that has been thought of/discussed/explored, since it was mainly designed for modules/plugins to both inform the user of specific behaviors in relation to the core engine and in the future allow core to 'inspect' these and react accordingly.  I really don't see us doing the latter for roles and think descriptions would be enough for this kind of thing. Even if I'm convinced this is a good feature going forward, I would not want to expand attributes use here in this PR, it should be with much more thought and input from the rest of the core team and a project onto itself.

felixfontein commented:
> Even if I'm convinced this is a good feature going forward, I would not want to expand attributes use here in this PR, it should be with much more thought and input from the rest of the core team and a project onto itself.

This PR is not expanding attributes. It is only fixing a bug introduced in another PR.

bcoca commented:
The bug (mine) was adding attributes to roles in the first place, that was never meant to be the case.

felixfontein commented:
BTW, ansible-lint already supports `seealso` and top-level `version_added`. https://github.com/ansible/ansible-lint/pull/4017 adds `notes` and `requirements`.

felixfontein commented:
I moved the attributes code to #82678. I will rebase with it removed from this PR.

felixfontein commented:
@bcoca what's the current state on this? Which parts of the PR are ok, which aren't?

s-hertel commented:
Separating out requirements into another PR like you did for attributes might help move this along. I agree that showing an aggregate list of requirements seems much more helpful than requiring double bookkeeping.

Not sure if I'm missing any relevant context, was there any discussion around why requirements was added in https://github.com/ansible-community/antsibull-docs/commit/87d3497ed987fbd28da2342d4abe5b3159792655? And why does ansibull-docs `RoleEntrypointSchema` have a `todo` field, is that another discrepancy?

I'm in favor of adding version_added/seealso/notes, especially the first two.

felixfontein committed:
Remove requirements.

felixfontein commented:
@bcoca @s-hertel I've updated the PR and removed `requirements` from it. The CI failures seem to be unrelated (`failed to fetch PPA information, error was: Connection failure: The read operation timed out`).

@s-hertel I guess you mean the schema added in https://github.com/ansible-community/antsibull-build/commit/f009a2f6fec8557c0bb9b7ae7c69f16d73a31713#diff-caae3b44cc89c5adfda1ef48009197d50f03c8ea9795aae7292b79cf7216d0d0R15 ? That was basically a copy of module/plugin documentation. Nobody commented on the `requirements` and `todo` fields back then, so there was no specific discussion about them. The PR was announced on May 28th, 2021 in The Bullhorn issue 27, asking for feedback; that was repeated in issue 28 ~a week later.

mattclay commented on lib/ansible/cli/doc.py:
```diff
@@ -1339,6 +1388,20 @@ def get_role_man_text(self, role, role_json):
                     text.append(DocCLI._indent_lines(DocCLI._dump_yaml(doc['attributes'][k]), opt_indent))
                 del doc['attributes']
 
+            if doc.get('notes', False):
+                text.append("")
+                text.append(_format("NOTES:", 'bold'))
+                for note in doc['notes']:
+                    text.append(DocCLI.warp_fill(DocCLI.tty_ify(note), limit - 6,
+                                                 initial_indent=opt_indent[:-2] + "* ", subsequent_indent=opt_indent))
+                del doc['notes']
+
+            if doc.get('seealso', False):
+                text.append("")
+                text.append(_format("SEE ALSO:", 'bold'))
+                DocCLI._add_seealso(text, doc['seealso'], limit=limit, opt_indent=opt_indent)
+                del doc['seealso']
```
This can be made less repetitive by using `doc.pop` and the walrus operator. Then there's only one reference to the key instead of three.

For example:

```python
        if notes := doc.pop('notes', None):
```

The same technique can be applied to the two other new conditionals. There's no need to update any of the existing code, just the portions new to this PR.

mattclay commented on lib/ansible/cli/doc.py:
```diff
@@ -1281,6 +1281,51 @@ def add_fields(text, fields, limit, opt_indent, return_values=False, base_indent
                 DocCLI.add_fields(text, subdata, limit, opt_indent + '  ', return_values, opt_indent)
 
+    @staticmethod
+    def _add_seealso(text, seealsos, limit, opt_indent):
```
I realize the body of this function is existing code that was relocated, but the function itself is new, so it would be good to have type annotations on the function declaration.

felixfontein committed:
Add typing info and use walrus operator.

mattclay commented on lib/ansible/cli/doc.py:
```diff
@@ -1308,6 +1354,10 @@ def get_role_man_text(self, role, role_json):
             text.append("ENTRY POINT: %s %s" % (_format(entry_point, "BOLD"), desc))
             text.append('')
 
+            if 'version_added' in doc:
```
This could be updated to use the walrus operator as well.

felixfontein committed:
One more walrus.

mattclay commented on lib/ansible/cli/doc.py:
```diff
@@ -1281,6 +1282,51 @@ def add_fields(text, fields, limit, opt_indent, return_values=False, base_indent
                 DocCLI.add_fields(text, subdata, limit, opt_indent + '  ', return_values, opt_indent)
 
+    @staticmethod
+    def _add_seealso(text: list[str], seealsos: list[dict[str, t.Any]], limit: int, opt_indent: str):
```
```suggestion
    def _add_seealso(text: list[str], seealsos: list[dict[str, t.Any]], limit: int, opt_indent: str) -> None:
```

felixfontein committed:
Add return type.

Co-authored-by: Matt Clay <matt@mystile.com>

felixfontein commented:
@bcoca can you merge this, or is this still waiting for something?

bcoca commented:
/azp run

felixfontein commented:
@bcoca @mattclay thanks for reviewing and merging!
